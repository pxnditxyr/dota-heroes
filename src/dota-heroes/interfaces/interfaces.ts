export interface Hero {
    id: string;
    name: string;
    another_name: string;
    attribute: string;
    attack: string;
    information: string;
};
