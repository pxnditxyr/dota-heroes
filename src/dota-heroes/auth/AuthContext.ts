import { createContext } from 'react';

import { User } from '../interfaces/interfaces';

export type AuthContextProps = {
    user: User;
    authLogin: ( user : User ) => void;
    authLogout: ( logged : boolean ) => void;
};

export const AuthContext = createContext<AuthContextProps>( {} as AuthContextProps );
