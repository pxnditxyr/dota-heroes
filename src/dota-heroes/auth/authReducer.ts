import { User } from '../interfaces/interfaces';

type authAction =
    | { type: "authLogin", payload: User }
    | { type: "authLogout", payload: { logged : boolean } }

export const authReducer = ( state : User, action : authAction ) : User => {

    switch ( action.type ) {
        case "authLogin":
            return {
                ...action.payload,
                logged: true,
            }

        case "authLogout":
            return {
                ...state,
                logged: false
            }
        default: 
            return state;
    }
};
