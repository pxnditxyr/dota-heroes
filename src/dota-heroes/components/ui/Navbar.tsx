import {
    NavLink,
} from 'react-router-dom';

export const Navbar = () => {
    return (
        <nav>
            <div>
                <NavLink
                    exact
                    to="/strenght"
                >
                    Strenght
                </NavLink>
                <NavLink
                    exact
                    to="/agility"
                >
                    Agility
                </NavLink>
                <NavLink
                    exact
                    to="/intelligence"
                >
                    Intelligence
                </NavLink>
            </div>
            <div>
                <NavLink
                    exact
                    to="/search"
                >
                    Search
                </NavLink>
                <NavLink
                    exact
                    to="/login"
                >
                    Logout
                </NavLink>
            </div>
        </nav>
    );
};
