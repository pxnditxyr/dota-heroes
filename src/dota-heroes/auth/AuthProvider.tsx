import { useEffect, useReducer } from 'react';
import { User } from '../interfaces/interfaces';

import { AuthContext } from './AuthContext';
import { authReducer } from './authReducer';


const INITIAL_STATE : User = {
    id: "",
    name: "",
    email: "",
    password: "",
    logged: false,
};

const init : () => User = () => {
    return JSON.parse( localStorage.getItem( 'user' ) || "{}" );
}

interface AuthProviderProps {
    children: JSX.Element | JSX.Element[];
};

export const AuthProvider = ( { children }: AuthProviderProps ) => {

    const [ user, dispatch ] = useReducer( authReducer, INITIAL_STATE, init );
    useEffect( () => {
        if ( !user ) return;
        localStorage.setItem( 'user', JSON.stringify( user ) );
    }, [ user ] )

    const authLogin = ( user: User ) => {
        dispatch({ type: "authLogin", payload: user });
    };

    const authLogout = ( logged : boolean ) => {
        dispatch({ type: "authLogout", payload: { logged } });
    };
    const auth = () => {
        dispatch({ type: "auth", payload: {} })
    }

    return (
        <AuthContext.Provider
            value={{
                user,
                authLogin,
                authLogout,
                auth,
            }}
        >
            { children }
        </AuthContext.Provider>
    );
};
