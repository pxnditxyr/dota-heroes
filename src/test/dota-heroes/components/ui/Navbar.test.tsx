import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router';
import { AuthContext } from '../../../../dota-heroes/auth/AuthContext';
import { Navbar } from '../../../../dota-heroes/components/ui/Navbar';
import { User } from '../../../../dota-heroes/interfaces/interfaces';

const mockNavigate = jest.fn();

jest.mock( 'react-router-dom', () => ({
    ...( jest.requireActual( 'react-router-dom' ) as object ),
    useNavigate: () => mockNavigate,
}));

describe( "Tests in <Navbar />", () => {
    test( "Must show correctly", () => {
        const contextValue = {
            user: {
                id: "1",
                name: "Yuki",
                email: "example@test.com",
                password: "MD5_test",
                logged: true,
            },
            authLogin: ( user : User ) => { console.log( user ) },
            authLogout: ( logged : boolean ) => { console.log( logged ) },
            auth: () => { console.log( "Test" ) },
        };
        const wrapper = mount(
            <AuthContext.Provider value={ contextValue }>
                <MemoryRouter initialEntries={[ '/' ]}>
                    <Navbar />
                </MemoryRouter>
            </AuthContext.Provider>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find( 'span' ).text().trim() ).toBe( contextValue.user.name );
    });

    test( "Must call logout, call navigate and authLogout with args", () => {
        
        const contextValue = {
            user: {
                id: "1",
                name: "Yuki",
                email: "example@test.com",
                password: "MD5_test",
                logged: true,
            },
            authLogin: jest.fn(),
            authLogout: jest.fn(),
            auth: jest.fn(),
        };
        const wrapper = mount(
            <AuthContext.Provider value={ contextValue }>
                <MemoryRouter initialEntries={[ '/' ]}>
                    <Navbar />
                </MemoryRouter>
            </AuthContext.Provider>
        );

        wrapper.find( 'button' ).simulate( 'click' );
        expect( mockNavigate ).toHaveBeenCalled();
        expect( mockNavigate ).toHaveBeenCalledWith( '/login' );
        expect( contextValue.authLogout ).toHaveBeenCalled();
        expect( contextValue.authLogout ).toHaveBeenCalledWith( false );
    });
});
