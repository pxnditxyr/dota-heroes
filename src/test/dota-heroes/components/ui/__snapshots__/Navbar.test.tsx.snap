// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <Navbar /> Must show correctly 1`] = `
<nav>
  <div>
    <a
      className=""
      href="/strenght"
      onClick={[Function]}
    >
      Strenght
    </a>
    <a
      className=""
      href="/agility"
      onClick={[Function]}
    >
      Agility
    </a>
    <a
      className=""
      href="/intelligence"
      onClick={[Function]}
    >
      Intelligence
    </a>
  </div>
  <div>
    <a
      className=""
      href="/search"
      onClick={[Function]}
    >
      Search
    </a>
    <span>
       
      Yuki
       
    </span>
    <button
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
</nav>
`;
